project device/xiaomi/sm8450-common/
diff --git a/BoardConfigCommon.mk b/BoardConfigCommon.mk
index 43d7ccd..aa3dc8c 100644
--- a/BoardConfigCommon.mk
+++ b/BoardConfigCommon.mk
@@ -220,7 +220,7 @@ endif
 BOARD_AVB_ENABLE := true
 BOARD_AVB_ALGORITHM := SHA256_RSA2048
 BOARD_AVB_KEY_PATH := external/avb/test/data/testkey_rsa2048.pem
-BOARD_AVB_MAKE_VBMETA_IMAGE_ARGS += --flags 3
+BOARD_AVB_MAKE_VBMETA_IMAGE_ARGS += --flags 0
 BOARD_AVB_RECOVERY_KEY_PATH := external/avb/test/data/testkey_rsa2048.pem
 BOARD_AVB_RECOVERY_ALGORITHM := SHA256_RSA2048
 BOARD_AVB_RECOVERY_ROLLBACK_INDEX := 1
