project kernel/xiaomi/sm8450/
diff --git a/arch/arm64/configs/vendor/debugfs.config b/arch/arm64/configs/vendor/debugfs.config
index d13694029..ebeaf37f9 100644
--- a/arch/arm64/configs/vendor/debugfs.config
+++ b/arch/arm64/configs/vendor/debugfs.config
@@ -1,4 +1,4 @@
-CONFIG_CMA_DEBUGFS=n
-CONFIG_PAGE_OWNER=n
-CONFIG_PAGE_PINNER=n
-CONFIG_DEBUG_FS=n
+CONFIG_CMA_DEBUGFS=y
+CONFIG_PAGE_OWNER=y
+CONFIG_PAGE_PINNER=y
+CONFIG_DEBUG_FS=y
diff --git a/arch/arm64/configs/vendor/xiaomi_GKI.config b/arch/arm64/configs/vendor/xiaomi_GKI.config
index 08f9ecde7..6169dae5d 100644
--- a/arch/arm64/configs/vendor/xiaomi_GKI.config
+++ b/arch/arm64/configs/vendor/xiaomi_GKI.config
@@ -58,7 +58,3 @@ CONFIG_TOUCHSCREEN_SYNAPTICS_TCM_SPI=m
 CONFIG_TOUCHSCREEN_SYNAPTICS_TCM_CORE=m
 CONFIG_TOUCHSCREEN_SYNAPTICS_TCM_REFLASH=m
 CONFIG_TOUCHSCREEN_XIAOMI_TOUCHFEATURE=m
-
-# USB
-CONFIG_USB_NET_AX88179_178A=m
-CONFIG_USB_NET_AX8817X=m
diff --git a/kernel/time/posix-cpu-timers.c b/kernel/time/posix-cpu-timers.c
index c8bc6722e..76f1cccb1 100644
--- a/kernel/time/posix-cpu-timers.c
+++ b/kernel/time/posix-cpu-timers.c
@@ -1373,6 +1373,9 @@ void run_posix_cpu_timers(void)
 
 	lockdep_assert_irqs_disabled();
 
+	if (tsk->exit_state)
+		return;
+
 	/*
 	 * Ensure that release_task(tsk) can't happen while
 	 * handle_posix_cpu_timers() is running. Otherwise, a concurrent
diff --git a/modules.list.msm.waipio b/modules.list.msm.waipio
index d4c5b8f2e..c58e430c0 100644
--- a/modules.list.msm.waipio
+++ b/modules.list.msm.waipio
@@ -5,6 +5,7 @@ clk-rpmh.ko
 gcc-waipio.ko
 icc-rpmh.ko
 qcom_ipcc.ko
+qcom_ipc_logging.ko
 qcom-pdc.ko
 qcom_tsens.ko
 rpmh-regulator.ko
@@ -21,6 +22,7 @@ cqhci.ko
 crypto-qti-common.ko
 crypto-qti-hwkm.ko
 dcvs_fp.ko
+debug-regulator.ko
 dispcc-diwali.ko
 dispcc-waipio.ko
 gcc-diwali.ko
@@ -78,6 +80,7 @@ qcom_wdt_core.ko
 qnoc-diwali.ko
 qnoc-qos.ko
 qrtr.ko
+qti-regmap-debugfs.ko
 regmap-spmi.ko
 rtc-pm8xxx.ko
 secure_buffer.ko


project device/xiaomi/sm8450-common/
diff --git a/BoardConfigCommon.mk b/BoardConfigCommon.mk
index 43d7ccd..aa3dc8c 100644
--- a/BoardConfigCommon.mk
+++ b/BoardConfigCommon.mk
@@ -68,6 +68,7 @@ BOARD_RAMDISK_USE_LZ4 := true
 BOARD_USES_GENERIC_KERNEL_IMAGE := true
 BOARD_USES_QCOM_MERGE_DTBS_SCRIPT := true
 TARGET_NEEDS_DTBOIMAGE := true
+TARGET_KERNEL_CLANG_VERSION := r416183b
 
 BOARD_KERNEL_BASE        := 0x00000000
 BOARD_KERNEL_PAGESIZE    := 4096
