project device/xiaomi/sm8450-common/
diff --git a/BoardConfigCommon.mk b/BoardConfigCommon.mk
index 43d7ccd..188de99 100644
--- a/BoardConfigCommon.mk
+++ b/BoardConfigCommon.mk
@@ -68,6 +68,7 @@ BOARD_RAMDISK_USE_LZ4 := true
 BOARD_USES_GENERIC_KERNEL_IMAGE := true
 BOARD_USES_QCOM_MERGE_DTBS_SCRIPT := true
 TARGET_NEEDS_DTBOIMAGE := true
+TARGET_KERNEL_CLANG_VERSION := r416183b
 
 BOARD_KERNEL_BASE        := 0x00000000
 BOARD_KERNEL_PAGESIZE    := 4096
@@ -147,12 +148,12 @@ BOARD_SUPER_PARTITION_GROUPS := qti_dynamic_partitions
 BOARD_QTI_DYNAMIC_PARTITIONS_PARTITION_LIST := odm product system system_ext vendor vendor_dlkm
 BOARD_QTI_DYNAMIC_PARTITIONS_SIZE := 9122611200 # 0x21FC00000 # BOARD_SUPER_PARTITION_SIZE - overhead (4MiB)
 
-BOARD_ODMIMAGE_FILE_SYSTEM_TYPE := ext4
-BOARD_PRODUCTIMAGE_FILE_SYSTEM_TYPE := ext4
-BOARD_SYSTEMIMAGE_FILE_SYSTEM_TYPE := ext4
-BOARD_SYSTEM_EXTIMAGE_FILE_SYSTEM_TYPE := ext4
-BOARD_VENDORIMAGE_FILE_SYSTEM_TYPE := ext4
-BOARD_VENDOR_DLKMIMAGE_FILE_SYSTEM_TYPE := ext4
+BOARD_ODMIMAGE_FILE_SYSTEM_TYPE := erofs
+BOARD_PRODUCTIMAGE_FILE_SYSTEM_TYPE := erofs
+BOARD_SYSTEMIMAGE_FILE_SYSTEM_TYPE := erofs
+BOARD_SYSTEM_EXTIMAGE_FILE_SYSTEM_TYPE := erofs
+BOARD_VENDORIMAGE_FILE_SYSTEM_TYPE := erofs
+BOARD_VENDOR_DLKMIMAGE_FILE_SYSTEM_TYPE := erofs
 
 TARGET_COPY_OUT_ODM := odm
 TARGET_COPY_OUT_PRODUCT := product
@@ -198,6 +199,8 @@ $(foreach sku, $(call to-upper, $(DEVICE_MANIFEST_SKUS)), \
     $(eval DEVICE_MANIFEST_$(sku)_FILES := \
         $(COMMON_PATH)/vintf/manifest.xml \
         $(COMMON_PATH)/vintf/manifest_xiaomi.xml \
+        $(COMMON_PATH)/vintf/vendor.dolby.media.c2.xml \
+        $(COMMON_PATH)/vintf/vendor.dolby.hardware.dms@2.0-service.xml \
         $(if $(TARGET_NFC_SUPPORTED_SKUS),$(COMMON_PATH)/vintf/manifest_no_nfc.xml,) \
     ))
 
@@ -208,6 +211,7 @@ $(foreach nfc_sku, $(call to-upper, $(TARGET_NFC_SUPPORTED_SKUS)), \
 endif
 
 DEVICE_FRAMEWORK_COMPATIBILITY_MATRIX_FILE += \
+    $(COMMON_PATH)/vintf/dolby_framework_matrix.xml \
     hardware/qcom-caf/common/vendor_framework_compatibility_matrix.xml \
     hardware/xiaomi/vintf/xiaomi_framework_compatibility_matrix.xml
 
@@ -220,7 +224,7 @@ endif
 BOARD_AVB_ENABLE := true
 BOARD_AVB_ALGORITHM := SHA256_RSA2048
 BOARD_AVB_KEY_PATH := external/avb/test/data/testkey_rsa2048.pem
-BOARD_AVB_MAKE_VBMETA_IMAGE_ARGS += --flags 3
+BOARD_AVB_MAKE_VBMETA_IMAGE_ARGS += --flags 0
 BOARD_AVB_RECOVERY_KEY_PATH := external/avb/test/data/testkey_rsa2048.pem
 BOARD_AVB_RECOVERY_ALGORITHM := SHA256_RSA2048
 BOARD_AVB_RECOVERY_ROLLBACK_INDEX := 1

project device/xiaomi/sm8450-common/
diff --git a/modules.list.second_stage b/modules.list.second_stage
index afc533f..9d5ef3a 100644
--- a/modules.list.second_stage
+++ b/modules.list.second_stage
@@ -28,6 +28,7 @@ videocc-diwali.ko
 bam_dma.ko
 gpi.ko
 pci-edma.ko
+qcom_cpuss_sleep_stats.ko
 mdt_loader.ko
 cdsp-loader.ko
 qcom_ramdump.ko
@@ -54,6 +55,7 @@ glink_pkt.ko
 glink_probe.ko
 pmic_glink.ko
 qti_battery_debug.ko
+charger-ulog-glink.ko
 altmode-glink.ko
 pmic-pon-log.ko
 msm_performance.ko
@@ -69,6 +71,7 @@ qcom_soc_wdt.ko
 sys_pm_vx.ko
 msm_ext_display.ko
 icnss2.ko
+rimps_log.ko
 plh_scmi.ko
 gplaf_scmi.ko
 shared_rail_scmi.ko
@@ -85,6 +88,7 @@ hvc_gunyah.ko
 msm_rng.ko
 frpc-adsprpc.ko
 rdbg.ko
+qcom_iommu_debug.ko
 lt9611uxc.ko
 msm_kgsl.ko
 zram.ko
@@ -103,8 +107,6 @@ wlan_firmware_service.ko
 cnss_plat_ipc_qmi_svc.ko
 cnss_nl.ko
 cnss_prealloc.ko
-asix.ko
-ax88179_178a.ko
 msm_sharedmem.ko
 phy-qcom-emu.ko
 phy-msm-ssusb-qmp.ko
@@ -118,6 +120,7 @@ usb_f_qdss.ko
 usb_f_ccid.ko
 usb_f_cdev.ko
 usb_f_gsi.ko
+f_fs_ipc_log.ko
 ucsi_glink.ko
 repeater.ko
 repeater-i2c-eusb2.ko
@@ -171,6 +174,7 @@ leds-qpnp-vibrator-ldo.ko
 plh_vendor.ko
 gplaf_vendor.ko
 shared_rail_vendor.ko
+tz_log.ko
 qce50.ko
 qcedev-mod.ko
 ipa_fmwk.ko
@@ -194,6 +198,7 @@ gh_virtio_backend.ko
 qcom-spmi-adc5.ko
 qcom-vadc-common.ko
 nvmem_qfprom.ko
+icc-test.ko
 qnoc-parrot.ko
 ofpart.ko
 chipreg.ko

project kernel/xiaomi/sm8450/
diff --git a/arch/arm64/configs/vendor/debugfs.config b/arch/arm64/configs/vendor/debugfs.config
index d13694029..ebeaf37f9 100644
--- a/arch/arm64/configs/vendor/debugfs.config
+++ b/arch/arm64/configs/vendor/debugfs.config
@@ -1,4 +1,4 @@
-CONFIG_CMA_DEBUGFS=n
-CONFIG_PAGE_OWNER=n
-CONFIG_PAGE_PINNER=n
-CONFIG_DEBUG_FS=n
+CONFIG_CMA_DEBUGFS=y
+CONFIG_PAGE_OWNER=y
+CONFIG_PAGE_PINNER=y
+CONFIG_DEBUG_FS=y

project kernel/xiaomi/sm8450/
diff --git a/arch/arm64/configs/vendor/xiaomi_GKI.config b/arch/arm64/configs/vendor/xiaomi_GKI.config
index 08f9ecde7..6169dae5d 100644
--- a/arch/arm64/configs/vendor/xiaomi_GKI.config
+++ b/arch/arm64/configs/vendor/xiaomi_GKI.config
@@ -58,7 +58,3 @@ CONFIG_TOUCHSCREEN_SYNAPTICS_TCM_SPI=m
 CONFIG_TOUCHSCREEN_SYNAPTICS_TCM_CORE=m
 CONFIG_TOUCHSCREEN_SYNAPTICS_TCM_REFLASH=m
 CONFIG_TOUCHSCREEN_XIAOMI_TOUCHFEATURE=m
-
-# USB
-CONFIG_USB_NET_AX88179_178A=m
-CONFIG_USB_NET_AX8817X=m

project kernel/xiaomi/sm8450/
diff --git a/kernel/time/posix-cpu-timers.c b/kernel/time/posix-cpu-timers.c
index c8bc6722e..76f1cccb1 100644
--- a/kernel/time/posix-cpu-timers.c
+++ b/kernel/time/posix-cpu-timers.c
@@ -1373,6 +1373,9 @@ void run_posix_cpu_timers(void)
 
 	lockdep_assert_irqs_disabled();
 
+	if (tsk->exit_state)
+		return;
+
 	/*
 	 * Ensure that release_task(tsk) can't happen while
 	 * handle_posix_cpu_timers() is running. Otherwise, a concurrent

project kernel/xiaomi/sm8450/
diff --git a/modules.list.msm.waipio b/modules.list.msm.waipio
index d4c5b8f2e..c58e430c0 100644
--- a/modules.list.msm.waipio
+++ b/modules.list.msm.waipio
@@ -5,6 +5,7 @@ clk-rpmh.ko
 gcc-waipio.ko
 icc-rpmh.ko
 qcom_ipcc.ko
+qcom_ipc_logging.ko
 qcom-pdc.ko
 qcom_tsens.ko
 rpmh-regulator.ko
@@ -21,6 +22,7 @@ cqhci.ko
 crypto-qti-common.ko
 crypto-qti-hwkm.ko
 dcvs_fp.ko
+debug-regulator.ko
 dispcc-diwali.ko
 dispcc-waipio.ko
 gcc-diwali.ko
@@ -78,6 +80,7 @@ qcom_wdt_core.ko
 qnoc-diwali.ko
 qnoc-qos.ko
 qrtr.ko
+qti-regmap-debugfs.ko
 regmap-spmi.ko
 rtc-pm8xxx.ko
 secure_buffer.ko


project prebuilts/clang/host/linux-x86/
diff --git a/.gitignore b/.gitignore
index 23f6953a..7273893c 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,2 +1,3 @@
 **/python3/**/*.pyc
 **/python3/**/__pycache__
+clang-r416183b/


